setup.py

```diff
@@ -1,4 +1,60 @@
-from setuptools import setup
+import sys
+
+try:
+    import _markerlib.markers
+except ImportError:
+    _markerlib = None
+
+try:
+    import pkg_resources
+except ImportError:
+    pkg_resources = None
+
+from setuptools import setup, __version__ as setuptools_version
+
+
+# _markerlib.default_environment() obtains its data from _VARS
+# and wraps it in another dict, but _markerlib_evaluate writes
+# to the dict while it is iterating the keys, causing an error
+# on Python 3 only.
+# Replace _markerlib.default_environment to return a custom dict
+# that has all the necessary markers, and ignores any writes.
+
+class Python3MarkerDict(dict):
+
+    def __setitem__(self, key, value):
+        pass
+
+    def pop(self, i=-1):
+        return self[i]
+
+
+if _markerlib and sys.version_info[0] == 3:
+    env = _markerlib.markers._VARS
+    for key in list(env.keys()):
+        new_key = key.replace('.', '_')
+        if new_key != key:
+            env[new_key] = env[key]
+
+    _markerlib.markers._VARS = Python3MarkerDict(env)
+
+    def default_environment():
+        return _markerlib.markers._VARS
+
+    _markerlib.default_environment = default_environment
+
+# Avoid the very buggy pkg_resources.parser, which doesnt consistently
+# recognise all markers until setuptools 20.10.0
+if pkg_resources:
+    if pkg_resources.parse_version(setuptools_version) < pkg_resources.parse_version('20.10.0'):
+        try:
+            MarkerEvaluation = pkg_resources.MarkerEvaluation
+
+            del pkg_resources.parser
+            pkg_resources.evaluate_marker = MarkerEvaluation._markerlib_evaluate
+            MarkerEvaluation.evaluate_marker = MarkerEvaluation._markerlib_evaluate
+        except AttributeError as e:
+            print('pkg_resources monkey patching', e)
```

Comment:
Yeah, I like to err on the side of no exception handling, to make the errors I don't expect loud.